jmx-monitoring/src/main/java/io/crate/beans/QueryStats.java

```diff
@@ -18,187 +18,159 @@
-import org.apache.logging.log4j.Logger;
-import org.elasticsearch.common.logging.Loggers;
-import org.elasticsearch.common.lucene.BytesRefs;
-import org.elasticsearch.common.settings.Settings;
-
-import javax.annotation.Nonnull;
-import java.util.*;
-import java.util.concurrent.CompletableFuture;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentMap;
-
-import static io.crate.action.sql.SQLOperations.Session.UNNAMED;
-import static io.crate.beans.QueryStats.MetricType.AVERAGE_DURATION;
-import static io.crate.beans.QueryStats.MetricType.FREQUENCY;
+import com.google.common.base.Supplier;
+import com.google.common.base.Suppliers;
+import io.crate.operation.collect.stats.JobsLogs;
+import io.crate.operation.reference.sys.job.JobContextLog;
+
+import java.util.HashMap;
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 
 public class QueryStats implements QueryStatsMBean {
 
-    private final Logger logger;
+    static class Commands {
+        static final String TOTAL = "total";
+        static final String UNCLASSIFIED = "unclassified";
 
-    enum MetricType {
-        FREQUENCY,
-        AVERAGE_DURATION
+        static final String SELECT = "select";
+        static final String INSERT = "insert";
+        static final String UPDATE = "update";
+        static final String DELETE = "delete";
+    }
+
+    static class Metric {
+
+        private final double elapsedSinceUpdateInSec;
+
+        private long count;
+        private long sumOfDurations;
+
+        Metric(long duration, double elapsedSinceUpdateInSec) {
+            this.elapsedSinceUpdateInSec = elapsedSinceUpdateInSec;
+            sumOfDurations = duration;
+            count = 1;
+        }
+
+        void inc(long duration) {
+            sumOfDurations += duration;
+            count++;
+        }
+
+        double statementsPerSec() {
+            return count / elapsedSinceUpdateInSec;
+        }
+
+        double avgDurationInMs() {
+            return sumOfDurations / count;
+        }
     }
 
     public static final String NAME = "io.crate.monitoring:type=QueryStats";
+    private static final Pattern COMMAND_PATTERN = Pattern.compile("^\\s*(select|insert|update|delete).*");
+    private static final Metric DEFAULT_METRIC = new Metric(0, 1);
+
+    private final Supplier<Map<String, Metric>> metricByCommand;
+
+    private volatile long lastUpdateTsInMillis = System.currentTimeMillis();
+
+    public QueryStats(JobsLogs jobsLogs) {
+        metricByCommand = Suppliers.memoizeWithExpiration(
+            () -> {
+                long currentTs = System.currentTimeMillis();
+                Map<String, Metric> metricByCommand = createMetricsMap(jobsLogs.jobsLog(), currentTs, lastUpdateTsInMillis);
+                lastUpdateTsInMillis = currentTs;
+                return metricByCommand;
+            },
+            1,
```

Comment:
This causes the metrics to be re-calculated at most once per sec - is this too low/too big?